### PR TITLE
Resource: Managed Solution

### LINKED ISSUE #1143: Resource: `powerplatform_managed_solution`
## User Story

As a Terraform user, I want to manage managed Power Platform solution deployment through Terraform so that I can promote versioned solution artifacts into environments with deterministic preflight validation and without relying on zip hashes as the deployment trigger.

Use case: A team publishes managed solution artifacts to a build output path or artifact URL. Terraform should deploy a specific `unique_name` + `version` into a target environment, verify that required connection references are bound, verify that referenced environment variables are already satisfiable in the target environment, verify required solution dependencies are already installed at sufficient versions, and then import the solution. If the environment is not ready, apply must fail before import.

## Resource

- **Resource Name:** `powerplatform_managed_solution`
- **Service Name:** `managed_solution`
- **Documentation Link:**
  - Existing solution resource: [solution.md](/Users/adam/Code/GitHub/AdamCoulterOz/terraform-provider-power-platform/docs/resources/solution.md)
  - PAC CLI import reference: [pac solution import](https://learn.microsoft.com/power-platform/developer/cli/reference/solution#pac-solution-import)
  - Deployment settings / connection refs / env vars: [Microsoft Learn](https://learn.microsoft.com/en-us/power-platform/alm/conn-ref-env-variables-build-tools)

### Potential Terraform Configuration

```hcl
resource "powerplatform_managed_solution" "core" {
  environment_id = powerplatform_environment.environment.id
  unique_name    = "contoso_core"
  version        = "1.2.3.4"

  source = {
    path = "${path.module}/contoso_core_1.2.3.4_managed.zip"
    # or
    # url = "https://artifacts.example.com/contoso_core_1.2.3.4_managed.zip"
  }

  connection_references = {
    contoso_sharedsharepoint = "00000000-0000-0000-0000-000000000000"
    contoso_sharedsql        = "11111111-1111-1111-1111-111111111111"
  }

  timeouts = {
    create = "30m"
    update = "30m"
    read   = "5m"
    delete = "30m"
  }
}
```

### Additional Validation Rules

- `environment_id` is required and changing it forces replacement.
- `unique_name` is required and changing it forces replacement.
- `version` is required and is the deployment trigger. The resource must not use zip checksum or file hash to determine drift.
- `source.path` xor `source.url` must be set.
- `source` changes alone do not trigger update if `unique_name` and `version` are unchanged.
- The package must be a managed solution. If not, fail apply.
- The package `UniqueName` must equal `unique_name`. If not, fail apply.
- The package `Version` must equal `version`. If not, fail apply.
- `connection_references` is `map(string)` where key = connection reference logical name and value = connection id.
- Every connection reference declared by the package must be present in `connection_references`.
- No extra `connection_references` keys may be supplied.
- For each configured connection id, the provider resolves the connection in the target environment and derives its connector API id. `connector_id` is not user input.
- If a configured connection id cannot be resolved, fail apply.
- Environment variables are not managed by this resource.
- The resource must inspect the package for referenced environment variable definitions.
- For each referenced environment variable, apply is allowed only if at least one is true:
  - the solution definition includes a default value
  - the target environment already has a current value
- If a referenced environment variable has neither a packaged default nor an existing current value in the environment, fail apply.
- Solution dependencies must be parsed from `solution.xml -> ImportExportXml.SolutionManifest.MissingDependencies.*.MissingDependency.Required.schemaName/solution -> Name (Version)`.
- If the same dependency appears multiple times, resolve the highest required version for that dependency name.
- Before import, verify each dependency exists in the target environment at a version equal to or higher than the resolved required version.
- If any dependency is missing or below the required version, fail apply.
- Dataverse must exist in the environment. If not, fail apply.
- No fallback behavior. Invalid, incomplete, or non-deployable inputs must fail loudly.

## API documentation

Implementation should reuse existing solution import patterns where appropriate and add the minimum additional queries needed to validate:
- target environment readiness
- installed solution versions
- configured connection resolution
- existing environment variable values

The exact API shapes should be determined during implementation and grounded in the existing provider patterns and Dataverse capabilities, rather than pre-committed in this proposal.

### JSON

```json
{
  "identity": {
    "environment_id": "00000000-0000-0000-0000-000000000001",
    "unique_name": "contoso_core",
    "version": "1.2.3.4"
  },
  "source": {
    "path_or_url": "artifact location"
  },
  "connection_references": {
    "contoso_sharedsharepoint": "00000000-0000-0000-0000-000000000000"
  },
  "apply_time_preflight": {
    "package_metadata": {
      "managed": true,
      "unique_name": "contoso_core",
      "version": "1.2.3.4"
    },
    "environment_variables": {
      "declared": [
        {
          "schema_name": "contoso_ApiBaseUrl",
          "has_packaged_default": true
        },
        {
          "schema_name": "contoso_TenantId",
          "has_packaged_default": false,
          "must_exist_in_environment": true
        }
      ]
    },
    "dependencies": {
      "resolved_minimum_versions": {
        "contoso_base": "2.4.0.0",
        "microsoft_portalbase": "1.0.0.0"
      }
    }
  }
}
```

## Resource Behavior Summary

- **Plan-time:** use `environment_id + unique_name + version` as desired identity. Do not download, hash, or inspect the artifact just to decide whether update is needed.
- **Apply-time:** fetch/read the artifact, parse `solution.xml`, inspect package contents, validate connection refs, validate env var satisfiability, validate dependency versions, then import.
- **Environment variables:** checked, not managed.
- **Connection references:** required bindings, managed as deployment inputs.
- **Dependencies:** parsed from package `MissingDependencies`, highest version per dependency name wins, target environment must satisfy equal-or-higher installed version.

## Definition of Done

- [ ] Data Transfer Objects (dtos) in `dto.go`
- [ ] Resource Model in `model.go`
- [ ] API Client functions in `api_{name}.go`
- [ ] Resource Implementation in `resource_{name}.go`
- [ ] Unit Tests in `resource_{name}_test.go` for Happy Path, Error conditions, boundry cases
- [ ] Acceptance Tests in `resource_{name}_test.go` for Happy Path
- [ ] Resource Added to `provider.go` and `provider_test.go`
- [ ] Example in the `/examples` folder
- [ ] Schema documented using `MarkdownDescription`
- [ ] Change log entry `changie new -k added`
- [ ] Run `make precommit` before PR

See the [contributing guide](/Users/adam/Code/GitHub/AdamCoulterOz/terraform-provider-power-platform/CONTRIBUTING.md) for more information about what's expected for contributions.

### PR DESCRIPTION
**Pull request by AdamCoulterOz**
## Description

Adds `powerplatform_managed_solution` — a resource modeling the desired state "this managed solution package/version is installed in this environment", per the design in #1143.

Closes #1143

### Behavior

- **Identity & trigger:** `environment_id + unique_name + version` define desired state. `version` is the deployment trigger — no zip checksums or file hashes are used for drift detection, so re-packaging the same version does not churn plans.
- **Source:** `source.path` xor `source.url` locate the artifact; source changes alone do not trigger an update.
- **Apply-time preflight (fail loudly, before import):** the package must be managed, and its `UniqueName`/`Version` must match the configured identity; every connection reference declared by the package must be bound in `connection_references`; referenced environment variables must be satisfiable (packaged default, existing target value, or supplied import component parameter); Dataverse must exist in the environment.
- **Dependency validation** is two-layered: a client-side preflight parses the package's `solution.xml` `MissingDependencies` (highest required version per dependency) and fails fast with an aggregated report before any upload, and the platform's own `StageSolution` validation results (missing dependencies, validation messages) are surfaced as apply errors as the authoritative check. Version comparison deliberately tolerates build-segment drift for non-built-in dependencies (declared builds float independently of installed builds, e.g. PCF support solutions).
- **Install vs upgrade:** initial installation uses `ImportSolutionAsync`; when the same managed solution is already installed at a lower version, the resource performs an in-place staged upgrade via `StageAndUpgradeAsync`. A same-version pre-existing install is adopted into state instead of erroring; a lower configured version fails apply (deployments must move forward).
- **Import start safety:** the import start is a non-idempotent mutation whose outcome can be ambiguous on transient HTTP failures, so it executes through a new `ExecuteWithoutRetry` API-client primitive and recovers by polling the async operation rather than replaying the request.
- **Environment variables: package-shape rules only, no configuration surface.** Definitions and defaults are solution owned; values are owned by the deployment context and set only by their own resource. The platform resolves a variable with no default and no value to null (verified empirically), so unresolved references are legal and there is no resolution gate. The resource enforces two rules: a package must not ship current values, and must not ship a definition that already exists unmanaged in the target environment (verified empirically: the import silently absorbs it into the managed layer and uninstall would then delete it).

### Deltas from the #1143 proposal

- No `environment_variables` attribute: values are set by `powerplatform_environment_variable_value`, ordered after the solution that ships the definition with an ordinary `depends_on`.
- `skip_product_update_dependencies` (default `true`) and `publish_all_customizations` (default `false`) expose the corresponding import behaviors.
- Dependency validation combines a client-side `solution.xml` preflight (fast, aggregated failure before upload) with the authoritative server-side `StageSolution` validation, and ignores build-segment drift for non-built-in dependencies (see above).

This implementation has been exercised extensively in our own deployment pipelines (repeated installs, staged upgrades, adoption and re-run scenarios) before being prepared for contribution.

### Relationship to existing resources

Complements `powerplatform_unmanaged_solution` (#1117 / #1118): managed and unmanaged solutions have fundamentally different lifecycle semantics (immutable versioned artifact vs open authoring container), so they are modeled as distinct resources — see the naming discussion in #1136.

## Checklist

- [x] DTOs in `dto.go`
- [x] Resource model in `models.go`
- [x] API client functions in `api_managed_solution.go`
- [x] Resource implementation in `resource_managed_solution.go`
- [x] Unit tests (happy path, error conditions, boundary cases: version decisions, package inspection, dependency validation, no-replay import start)
- [x] Acceptance test (happy path)
- [x] Registered in `provider.go` and `provider_test.go`
- [x] Example in `/examples/resources/powerplatform_managed_solution`
- [x] Schema documented via `MarkdownDescription` (+ `docs/resources/managed_solution.md`)
- [x] Changelog entry (`.changes/unreleased`)
- [x] `go build` / lint / unit tests clean locally

### CODE REVIEW

File: internal/services/managed_solution/api_managed_solution.go [hunk 1]
```diff
@@ -0,0 +1,781 @@
...
+	if err := client.Api.HandleForbiddenResponse(resp); err != nil {
+		return err
+	}
+	if err := client.Api.HandleNotFoundResponse(resp); err != nil {
+		return err
+	}
+	if response.SolutionOperationResult.Status != "Passed" {
+		return fmt.Errorf("solution import failed: %s", response.SolutionOperationResult.ErrorMessages...)
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/managed_solution/api_managed_solution.go:226]: Fixed in c4b37c6e: the messages are now converted with `fmt.Sprint` and joined with `; ` into a single readable error string, so empty and multi-message results both format cleanly.

File: internal/services/managed_solution/resource_managed_solution.go [hunk 2]
```diff
@@ -0,0 +1,693 @@
...
+	}
+
+	installedSolutions, err := r.Client.GetInstalledSolutions(ctx, plan.EnvironmentId.ValueString())
+	if err != nil {
+		diagnostics.AddError("Unable to verify installed solution dependencies", err.Error())
+		return nil
+	}
+	if err := validateDependencies(pkg.Dependencies, installedSolutions); err != nil {
+		diagnostics.AddError("Managed solution dependency validation failed", err.Error())
+		return nil
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/managed_solution/resource_managed_solution.go:501]: Good catch - the PR description was inaccurate and has been updated. The implementation is deliberately two-layered: a client-side preflight parses the package's `solution.xml` `MissingDependencies` (highest required version per dependency) so apply fails fast with one aggregated report before any upload, and the platform's own `StageSolution` validation results remain the authoritative check surfaced as apply errors.

File: internal/services/managed_solution/api_managed_solution.go [hunk 3]
```diff
@@ -0,0 +1,781 @@
...
+			continue
+		}
+
+		cmp, err := compareVersionStringsIgnoringBuild(installedSolution.Version, requiredVersion)
+		if err != nil {
+			return err
+		}
+		if cmp < 0 {
+			problems = append(problems, fmt.Sprintf("required dependency %q >= %s but %s is installed", dependencyName, requiredVersion, installedSolution.Version))
+		}
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/managed_solution/api_managed_solution.go:690]: This tolerance is deliberate, and it is now documented on `compareVersionStringsIgnoringBuild` (c4b37c6e). Dependency requirements captured from a package's `MissingDependencies` pin a build number that floats independently of the installed dependency's build - PCF support solutions being the canonical case - so requiring an equal-or-higher build fails deployments that are actually compatible. Major.minor.patch is still enforced as equal-or-higher, and the behavior is locked by `TestUnitValidateDependencies_AllowsLowerBuildWhenMajorMinorPatchMatch`.

**Comment by AdamCoulterOz** [comment]
Pushed the environment-variable reference-binding model per the disposition analysis on #1136 (§C): `environment_variables` now binds each schema name to the owning `powerplatform_environment_variable` resource id, creating a dependency-ordering edge and a preflight assertion that every package-declared reference resolves before upload. The resource never transmits or writes environment-variable values (packaged current values were already rejected at preflight). Also switched the composite id / import format to the provider-conventional `{environment_id}/{solution_id}`. Thanks for the example and acceptance-test fixes @mawasile — merged in cleanly.

**Comment by mawasile** [comment]
@AdamCoulterOz please merge https://github.com/AdamCoulterOz/terraform-provider-power-platform/pull/4 PR

**Review by mawasile — DISMISSED** [review]

**Comment by AdamCoulterOz** [comment]
Pushed the simplified environment variable model (26407391), following the empirical work described: the platform returns 200 with a null value for a variable with no default and no value, so unresolved references are legal and the resolution gate is gone. The environment_variables attribute is removed entirely; values belong to their own resource, ordered with depends_on. What remains is package-shape policing: no packaged current values, and no shipping a definition that already exists unmanaged in the target (the import silently absorbs it into the managed layer, so preflight rejects that collision). Your PR4 example now works with no changes to the solution zips; only the config map came out of the example. Tests and lint green.

File: internal/services/managed_solution/resource_managed_solution.go [hunk 4]
```diff
@@ -0,0 +1,693 @@
...
+				Required:            true,
+				Attributes: map[string]schema.Attribute{
+					"path": schema.StringAttribute{
+						MarkdownDescription: "Local filesystem path to the managed solution zip package.",
+						Optional:            true,
+					},
+					"url": schema.StringAttribute{
+						MarkdownDescription: "Remote URL to the managed solution zip package.",
+						Optional:            true,
+					},
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/managed_solution/resource_managed_solution.go:105]: Fixed in c4b37c6e: source.url is now Sensitive, with the schema description noting artifact URLs commonly embed SAS tokens.

File: .changes/unreleased/added-20260409-161500.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+kind: added
+body: Add `powerplatform_managed_solution` resource for identity-driven managed solution deployment with apply-time package validation for managed state, connection references, environment variable satisfiability, and dependency versions
+time: 2026-04-09T16:15:00Z
+custom:
+    Issue: ""
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply .changes/unreleased/added-20260409-161500.yaml]: Fixed in c4b37c6e: the changelog entry now carries Issue "1143".

**Review by ianjensenisme — APPROVED** [review]
